### ISSUE TITLE
Twisted Names gethostbyname.py example fails to exit if the host is in /etc/hosts

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5596]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5596|
|Type|defect|
|Created|2012-04-02 15:20:15Z|
|Branch|https://github.com/twisted/twisted/tree/gethostbyname-example-5596|

The /etc/hosts resolver operates synchronously, so the `gotResult` callback in this example runs before `reactor.run`.  The `reactor.stop` call it contains therefore fails, and there isn't even an errback to handle that failure.  The example appears to work, by reporting the address of the host, but then does not exit and does not explain why it is not exiting (ie, reports no error).  At least until you hit !^C, and then it displays a traceback and exits.

Should be pretty easily reproducable - just try it on `localhost`.

This is an example of a very common bug.  [#11551](https://github.com/twisted/twisted/issues/11551) and [#3270](https://github.com/twisted/twisted/issues/3270) both aim to solve the bug more generally, so they may be applicable here somehow.

Attachments:

* [gethostbyname-exit-5596.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/08b/08bf6ca3fcdcb44b72fd44694748fab054dab19f/8381f16ccfd67532c4aeec156cbbcefdeec9795d.patch) (1535 bytes) - added by rwall on 2012-09-09 11:31:50Z - A patch which postpones the lookup until the reactor has started
* [gethostbyname-exit-5596-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/08b/08bf6ca3fcdcb44b72fd44694748fab054dab19f/490a5b5a9c3596f8d86132059204563244d3fa6a.patch) (2957 bytes) - added by rwall on 2012-09-14 02:37:27Z - added better error reporting
* [gethostbyname-exit-5596-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/08b/08bf6ca3fcdcb44b72fd44694748fab054dab19f/e906d4477fcc2f2ccfb98401e6d8317b4b5c865d.patch) (2216 bytes) - added by rwall on 2012-09-15 16:54:36Z - Updated to use task.react
* [gethostbyname-exit-5596-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/08b/08bf6ca3fcdcb44b72fd44694748fab054dab19f/786f12660adcfb197cee0610a57014dbe41f84af.patch) (9820 bytes) - added by rwall on 2012-09-15 19:50:07Z - Updated the other names examples to use task.react
* [names-examples-tests-5596.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/08b/08bf6ca3fcdcb44b72fd44694748fab054dab19f/80ec0fa011874e363320622686f92ef8f49ee733.patch) (4061 bytes) - added by rwall on 2012-09-16 02:20:00Z - Some basic tests for these examples

<details><summary>Searchable metadata</summary>

```
trac-id__5596 5596
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_gethostbyname_example_5596 branches/gethostbyname-example-5596
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1333380015000000 1333380015000000
changetime__1356816870000000 1356816870000000
version__None None
owner__exarkun exarkun
cc__rwall cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Yep. The problem is easily recreated.

```
[richard@zorin trunk]$ python doc/names/examples/gethostbyname.py localhost
127.0.0.1
^CUnhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
  File "doc/names/examples/gethostbyname.py", line 26, in <module>
    d.addCallbacks(gotResult, gotFailure)
  File "/home/richard/projects/Twisted/trunk/twisted/internet/defer.py", line 290, in addCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/home/richard/projects/Twisted/trunk/twisted/internet/defer.py", line 551, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "doc/names/examples/gethostbyname.py", line 19, in gotResult
    reactor.stop()
  File "/home/richard/projects/Twisted/trunk/twisted/internet/base.py", line 577, in stop
    "Can't stop reactor that isn't running.")
twisted.internet.error.ReactorNotRunning: Can't stop reactor that isn't running.

```

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Ready for review in [gethostbyname-exit-5596.patch](None/gethostbyname-exit-5596.patch)
 * Call getHostByName via reactor.callWhenRunning so that reactor is started and stopped in the correct order.
 * Added some doc comments explaining how the default resolver works and explaining that hosts lookup is synchronous.

Output now looks like this:
```
[richard@zorin gethostbyname-exit-5596]$ python doc/names/examples/gethostbyname.py localhost
127.0.0.1

[richard@zorin gethostbyname-exit-5596]$ python doc/names/examples/gethostbyname.py www.example.com
192.0.43.10

[richard@zorin gethostbyname-exit-5596]$ python doc/names/examples/gethostbyname.py does.not.exist
Traceback (most recent call last):
Failure: twisted.names.error.DNSNameError: <twisted.names.dns.Message instance at 0x2a422d8>

```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra set owner to @wallrj|
|-|-|

Replying to [rwall](#note_2):
> [richard@zorin gethostbyname-exit-5596]$ python doc/names/examples/gethostbyname.py does.not.exist
> Traceback (most recent call last):
> Failure: twisted.names.error.DNSNameError: <twisted.names.dns.Message instance at 0x2a422d8>
>

Thanks for the patch. Is it possible to print more details for that `Failure: twisted.names.error.DNSNameError: <twisted.names.dns.Message instance at 0x2a422d8>`? I assume someone will want to see more details when encountering a failure..

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj commented|
|-|-|

Replying to [thijs](#note_3):
> Thanks for the patch. Is it possible to print more details for that `Failure: twisted.names.error.DNSNameError: <twisted.names.dns.Message instance at 0x2a422d8>`? I assume someone will want to see more details when encountering a failure.

Thanks for the review. I agree and have added better error messages.

Have a look at [gethostbyname-exit-5596-2.patch](None/gethostbyname-exit-5596-2.patch)

Output now looks like this:

```
$ python gethostbyname.py www.google.com
2a00:1450:400c:c05::67
$ python gethostbyname.py www.google.com
173.194.67.104
$ python gethostbyname.py localhost
127.0.0.1
$ python gethostbyname.py com
ERROR: No IP adresses found for hostname 'com'
$ python gethostbyname.py does.not.exist
ERROR: hostname not found 'does.not.exist'
```

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

See [gethostbyname-exit-5596-3.patch](None/gethostbyname-exit-5596-3.patch)
 * Updated again to use exarkun's new task.react function from [#3270](https://github.com/twisted/twisted/issues/3270)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj commented|
|-|-|

Another patch [gethostbyname-exit-5596-4.patch](None/gethostbyname-exit-5596-4.patch)
 * Use task.react in all the names example scripts
 * use consistent names and documentation for all the names examples
 * Tidied up the output and error reporting of dns-service.py

Also fixes: [#5989](https://github.com/twisted/twisted/issues/5989)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj commented|
|-|-|

In [names-examples-tests-5596.patch](None/names-examples-tests-5596.patch)

 * Added tests for consistent USAGE documentation and command line error reporting
 * Added tests for standard python shebang line and that the script is executable.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [36679]) Branching to 'gethostbyname-example-5596'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [36680]) Apply part of gethostbyname-exit-5596-4.patch 

refs [#5596](https://github.com/twisted/twisted/issues/5596)

Drop all the parts not related to the ticket description

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  To simplify the review, I only considered the part of the patch which fixes the problem described by the ticket summary.

That parts looks great.  I'll apply it to trunk.  The other parts should be attached to new (or existing) tickets.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [36682]) Merge gethostbyname-example-5596

Author: rwall
Reviewer: exarkun
Fixes: [#5596](https://github.com/twisted/twisted/issues/5596)

Reliably exit the `gethostbyname.py` Twisted Names example by using `twisted.internet.task.react`
to run and stop the reactor, rather than doing it directly in an error prone manner.
